kubernetes-client-api/src/main/java/io/fabric8/kubernetes/client/utils/HttpClientUtils.java

```diff
@@ -73,7 +74,6 @@ public void before(BasicBuilder builder, HttpRequest request, RequestTags tags)
   private static final Pattern IPV4_PATTERN = Pattern.compile(
       "(http://|https://)?(?<ipAddressOrSubnet>(([01]?\\d\\d?|2[0-4]\\d|25[0-5])\\.){3}([01]?\\d\\d?|2[0-4]\\d|25[0-5])(\\/([1-2]\\d|3[0-2]|\\d))?)(\\D+|$)");
   private static final Pattern INVALID_HOST_PATTERN = Pattern.compile("[^\\da-zA-Z.\\-/:]+");
-  private static final AtomicBoolean MULTIPLE_HTTP_CLIENT_WARNING_LOGGED = new AtomicBoolean();
```

Comment:
This could get a little chatty in some extreme case - the user has multiple http clients in the classpath, relies upon the default logic to create one of the client types, but specifies the other one programmatically.

Reply:
Shouldn't they all get picked up by the ServiceLoader at once, though?

Reply:
Or is it a matter that the method might be called several times, thus leading to the warning being output several times? If that's the case, this actually raises the question of whether or not we should record the chosen factory "once and for all" to avoid going through the resolution process every time a new client is created? Another potential issue is whether or not we want to support creating clients with implicitly different http clients?

Reply:
> Or is it a matter that the method might be called several times, thus leading to the warning being output several times?

Yes, every client KubernetesClient creation that does not specify a HttpClient.Factory makes this call.

> If that's the case, this actually raises the question of whether or not we should record the chosen factory "once and for all" to avoid going through the resolution process every time a new client is created? 

Possibly.  Some people would cache the ServiceLoader, it's iterator will provide new instances should they be found later.  However I would not want this to be static as the current HttpClientUtils is.  There are several other usages of ServiceLoader in fabric8 that are created for each use.  The only one that is not I believe is the KubernetesDeserializer because the resulting mapping is held statically.

> Another potential issue is whether or not we want to support creating clients with implicitly different http clients?

That seems fine if your circumstances allow for that - you'd have to be doing something odd like dynamically installing new services or changing the priority of an existing one for the selection to change.

Reply:
I have to admit that's a little concerning because I was under the (maybe misguided) impression that creating a client was a rather fast operation. If we do http client resolution each time now, i.e. class path scanning, it might not be as much. If installing new services dynamically is not considered a core use case, it should be fine to always use the same resolved instance all the time (after the initial lookup).

Reply:
> (maybe misguided) impression that creating a client was a rather fast operation

It should be especially in comparison to how long it should be used - any remote round trip will end up being orders of magnitude greater than the time to create the client.

>  If we do http client resolution each time now, i.e. class path scanning, it might not be as much

The ServiceToURLProvider has the same approach and you could make the same case about places that make repeated calls to loadClass.

> If installing new services dynamically is not considered a core use case, it should be fine to always use the same resolved instance all the time (after the initial lookup)

It's mostly a question of where to put it - if you want to refactor so that it's held on the KuberenetesClientBuilder in some fashion that would be fine, but I would avoid doing anything statically.